Prox/Prox/Data/PlacesController.swift

```diff
@@ -0,0 +1,104 @@
+}
+
+private let apiSuffix = "/api/v1.0/at/%f/%f"
+
+class PlacesController {
+    weak var delegate: PlacesControllerDelegate?
+
+    private let database = FirebasePlacesDatabase()
+
+    private var isUpdating = false
+
+    private lazy var sessionManager: AFHTTPSessionManager = {
+        let manager = AFHTTPSessionManager(sessionConfiguration: .default)
+        manager.responseSerializer = AFHTTPResponseSerializer()
+        return manager
+    }()
+
+    func updatePlaces(forLocation location: CLLocation) {
+        assert(Thread.isMainThread)
+        // We would like to prevent running more than one update at the same time.
+        if isUpdating {
+            return
+        }
+        isUpdating = true
+
+        self.delegate?.placeControllerWillStartFetchingPlaces(self)
+
+        // Tell the server that we're here. We don't need to do anything with the return
+        // value: we're just instructing the server to go crawl the venues/events the area 
+        // we're in now.
+        let coord = location.coordinate
+        let path  = String(format:apiSuffix, coord.latitude, coord.longitude)
+        let url   = AppConstants.serverURL.appendingPathComponent(path)
+
+        sessionManager.put(url.absoluteString, parameters: nil,
+            success: { (task, data) in
+                // TODO is there anything useful we can get from the server to help the UI here?
+                // e.g. some JSON to give queue size (how busy is it).
+                print("Server responded ok.")
+            },
+            failure: { (task, err) in
+                print("Error from server: \(err)")
+                DispatchQueue.main.async {
+                    self.delegate?.placesController(self, didError: err)
+                }
+            }
+        )
+
+        // Immediately after we've told the server, we should start querying the firebase datastore,
+        // because we may have already got something cached.
+        retryQueryPlaces(location: location, retriesLeft: 60, lastCount: 0)
+    }
+
+    private func retryQueryPlaces(location: CLLocation, retriesLeft: Int, lastCount: Int) {
+        // Fetch a stable list of places from firebase.
+        // In the event of the server crawling (from a cold start, for example)
+        // the server will be adding places to firebase.
+        // We want to wait for the number of firebase results to stop changing.
+        database.getPlaces(forLocation: location).upon { results in
+            let places = results.flatMap { $0.successResult() }
+            let placeCount = places.count
+            // Check if we have a stable number of places.
+            if (placeCount > 0 && lastCount == placeCount) || retriesLeft == 0 {
+                let sortedPlaces = self.preparePlaces(places: places, forLocation: location)
+                DispatchQueue.main.async {
+                    self.delegate?.placesController(self, didReceivePlaces: sortedPlaces)
```

Comment:
it would be nice if this happened in every recursion, rather than just at the end. That way we can start displaying places as soon as we start getting them, rather than only when we have fetched them all. It would make the wait time for full results less obvious. But maybe that is an enhancement for later